apis/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,169 @@
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified,
+	// this field defaults to TCP.
+	// +kubebuilder:default=TCP
+	Protocol *v1.Protocol `json:"protocol"`
+
+	// Number defines a network port value.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Number int32 `json:"number"`
+}
+
+// PortRange defines an inclusive range of ports from the the assigned Start value
+// to End value.
+type PortRange struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified,
+	// this field defaults to TCP.
+	// +kubebuilder:default=TCP
+	Protocol *v1.Protocol `json:"protocol,omitempty"`
+
+	// Start defines a network port that is the start of a port range, the Start
+	// value must be less than End.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Start int32 `json:"start"`
+
+	// End defines a network port that is the end of a port range, the End value
+	// must be greater than Start.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	End int32 `json:"end"`
+}
+
+// AdminNetworkPolicyPeer defines an in-cluster peer to allow traffic to/from.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyPeer struct {
+	// Namespaces defines a way to select a set of Namespaces.
+	// +optional
+	Namespaces *NamespacedPeer `json:"namespaces,omitempty"`
+	// NamespacedPods defines a way to select a set of pods in
+	// in a set of namespaces.
+	// +optional
+	Pods *NamespacedPodPeer `json:"pods,omitempty"`
+}
+
+type NamespaceRelation string
+
+const (
+	NamespaceSame    NamespaceRelation = "Self"
+	NamespaceNotSame NamespaceRelation = "NotSelf"
+)
+
+// NamespacedPeer defines a flexible way to select Namespaces in a cluster.
+// Exactly one of the selectors must be set.  If a consumer observes none of
+// its fields are set, they must assume an unknown option has been specified
+// and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type NamespacedPeer struct {
+	// Related provides a mechanism for selecting namespaces. It can be set to
+	// "Self" to select all pods in the subject's namespace, and set to "NotSelf"
+	// to select all pods not in the subject's namespace.
+	// +optional
+	Related *NamespaceRelation `json:"related,omitempty"`
+
+	// NamespaceSelector is a labelSelector used to select Namespaces, This field
+	// follows standard label selector semantics; if present but empty, it selects
+	// all Namespaces.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+	// SameLabels is used to select a set of Namespaces that share the same values
```

Comment:
hm... it's too bad we managed to squash Self/NotSelf but didn't manage to squash SameLabels/NotSameLabels.

It's also too bad that one of them uses "Self" and the other uses "Same" (which is confusing enough that you already got it wrong, above).

I don't have any suggestions for v1alpha1, but I think this might be something we could improve later...

Reply:
Yeah I'll think on this a bit more, We probably could combine into a `RelatedLabels` or something of the sort with a list of labels and a single Enum shared with `Related` possibly